Library/ViewModels/SurveyResponseViewModel.swift

```diff
@@ -64,37 +57,59 @@ public final class SurveyResponseViewModel: SurveyResponseViewModelType {
       }
       .skipNil()
 
-    let requestAndNavigationType = self.policyForNavigationActionProperty.signal.skipNil()
-      .map { action in (action.request, action.navigationType) }
+    let newRequest = self.policyForNavigationActionProperty.signal.skipNil()
+      .map { action in action.request }
 
-    let surveyPostRequest = requestAndNavigationType
-      .filter { request, navigationType in
-        isUnpreparedSurvey(request: request) && navigationType == .formSubmitted
+    let newSurveyRequest = newRequest
+      .filter { request in
+        isUnpreparedSurvey(request: request)
       }
-      .map { request, _ in request }
 
-    let redirectAfterPostRequest = requestAndNavigationType
-      .filter { request, navigationType in
-        isUnpreparedSurvey(request: request) && navigationType == .other
-      }
-      .map { request, _ in request }
-
-    self.dismissViewController = Signal.merge(
-      self.alertButtonTappedProperty.signal,
-      self.closeButtonTappedProperty.signal
-    )
+    self.dismissViewController = self.closeButtonTappedProperty.signal
 
-    self.goToProject = requestAndNavigationType
-      .map { request, _ -> (Param, RefTag?)? in
+    self.goToProject = newRequest
+      .map { request -> (Param, RefTag?)? in
         if case let (.project(param, .root, refInfo))? = Navigation.match(request) {
           return (param, refInfo?.refTag)
         }
         return nil
       }
       .skipNil()
 
-    self.policyDecisionProperty <~ requestAndNavigationType
-      .map { request, _ in
+    self.goToPledge = newRequest
+      .map { request -> (Param)? in
+        if case let (.project(param, .pledge, refInfo))? = Navigation.match(request) {
+          return param
+        }
+        return nil
+      }
+      .skipNil()
+
+    self.goToUpdate = newRequest
+      .map { (request: URLRequest) -> (Param, Int)? in
+        if case let (.project(param, .update(id, _), _))? = Navigation.match(request) {
+          return (param, id)
+        }
+        return nil
+      }
+      .skipNil()
+      .switchMap { (param: Param, updateId: Int) in
+        AppEnvironment.current.apiService.fetchProject(param: param)
+          .demoteErrors()
+          .map { project -> (Param, Project, Int) in
+            (param, project, updateId)
+          }
+      }
+      .switchMap { (param: Param, project: Project, updateId: Int) in
+        AppEnvironment.current.apiService.fetchUpdate(updateId: updateId, projectParam: param)
+          .demoteErrors()
+          .map { update -> (Project, Update) in
+            (project, update)
+          }
+      }
```

Comment:
Nit: will this show a spinner or some other UI while it makes two requests to the Kickstarter backend? Seems like it might be slow.

Reply:
That's a good point but I can't make the update page responsible for this unless I refactor the UpdateViewModel (very doable) and the ShareViewModel (looks kind of terrible, since the Project and Update are associated types for the config enum)... And adding a loading indicator between two webviews is not something I know how to do. I'll file a followup ticket to refactor the UpdateViewModel and ShareViewModel to make this possible.

Reply:
Solid, that makes sense.

Reply:
Ticket here: https://kickstarter.atlassian.net/browse/MBL-1533 for reference